### ISSUE TITLE
iOS foreground push notifications not displaying

### ISSUE THREAD

**Issue by kevjose** [ben-xD/push#4]
Hey @ben-xD, thank you for this amazing package.
I was testing the package out and noticed that the push notifications were not working on iOS. While a push notification in the foreground is caught it throws an error
```
userNotificationCenter:willPresent: received error: nil
```
On investigating further, I found that if I change the `guard` clause  at https://github.com/ben-xD/push/blob/main/push_ios/ios/Classes/UserNotificationCenterDelegateHandlers.swift#L16 to `guard error == nil else {`, the foreground messages start to work 

I am not very well versed with swift, hence a bit skeptical about raising a PR for this. Could you confirm that this is a valid bug and that the above solution is correct?

**Comment by ben-xD** [comment]
Hey @kevjose, 

Thanks for reporting the issue 👍. It's a little more complicated than just the guard statement - though that itself is a bug.

TLDR: I've added a checkbox ("Show foreground notifications") to the example app to show foreground notifications. If you want any notifications to be shown, you can enable it. It just uses `flutter_local_notifications` to do this. It took a while to fix it this evening because there was a bug exactly where you pointed :). 
![screenshot](https://user-images.githubusercontent.com/24711048/168179457-ebb50463-1111-4fe8-a0c4-0597abf5ba21.png)

## My original intention

I intentionally avoid showing foreground notifications by default to be consistent with Android. I had documented it in the README:

>  Notifications are not shown to the user **when the app is in the foreground**. If you want to show the user a notification, you should send a background notification to your app, and **create a local notification**.

I did that because the default behaviour on Android is to hide foreground push notifications - in fact, there is no way to change that directly. I wanted the API to be consistent between Android and iOS.

## A bug 

If you tried to display foreground notifications with `flutter_local_notifications`, it would not work. I was also ignoring those local notifications. I've now fixed it.

Since this is a common use-case, I've added this option to the example.

## Random comment

The federated plugin architecture makes changes like this which affect multiple federated packages quite tedious.

## Thanks for trying the package and reporting the issue

I'm going to make the release now. LMK if you have other issues or questions :)

**Comment by ben-xD** [comment]
I've released the fix and example app improvement in https://pub.dev/packages/push/versions/0.1.1

Thanks @kevjose 👍

**Comment by kevjose** [comment]
@ben-xD Thank you for fixing this and the detailed explanation.

I had gone through the doc where you had pointed out about the default i.e no foreground notifications being shown, however, the function `Push.instance.onShowNotificationInForegroundHandler` got me confused, as I thought that it would override the default.

I appreciate your work here :)

**Comment by ben-xD** [comment]
That confusing `onShowNotificationInForegroundHandler` was leftover code from when I did intend to allow apps to configure foreground notifications using `push` - then I avoided showing foreground push notifications on Android in this package. I would have needed to "create" a push notification manually on the Android side in the cases where foreground push notifications was enabled. 

This won't be very customizable if I manually create this on Android compared to using `flutter_local_notifications`.

**Comment by kevjose** [comment]
@ben-xD I tested the latest version 0.1.2, however, `push.Push.instance.onMessage` is only triggered on Android and does not trigger on iOS at all
(could it be that the screenshot you have posted above is from an android emulator ?)

**Comment by ben-xD** [comment]
Hey @kevjose, that screenshot is on a real iOS device. I've tested it again, and it still works. 

Here's a recording: 

https://user-images.githubusercontent.com/24711048/168281812-9da31f1c-7566-46be-9cd4-5cd24c42fae1.mp4

**Comment by kevjose** [comment]
@ben-xD I actually do receive the notification from APNs and I was able to check that by printing it at, https://github.com/ben-xD/push/blob/main/push_ios/ios/Classes/UserNotificationCenterDelegateHandlers.swift#L18 however, once it is ignored it is not being caught in the `onMessage` handler and further letting `flutter_local_notifications` create a local push notification.

I will fork your example app and try to replicate the exact same setup and figure out the issue.

**Comment by ben-xD** [comment]
I have not tested it with OneSignal, but I just tried. Unfortunately, I got this error multiple times:
![onesignal](https://user-images.githubusercontent.com/24711048/168380750-6b2f6241-b5b7-441a-9cd6-82b3d7f18ff5.png)

**Comment by ben-xD** [comment]
Okay I reproduced it, and I'll try to figure out what's going on :)

**Comment by ben-xD** [comment]
TLDR: This is an incompatibility between **onesignal_flutter** and **flutter_local_notifications**

Merely having OneSignal installed causes some of its initialization code to run. This breaks the `flutter_local_notification` plugin, because for some reason, `userNotificationCenter_willPresent` is not called - this means local notifications do not work when OneSignal is installed.

Then I found https://github.com/MaikuB/flutter_local_notifications/issues/440. Someone suggested a potential solution - I tried it but it didn't work.

It requires a bit more investigation on OneSignal's interoperability with other packages. The root cause is probably in https://github.com/OneSignal/OneSignal-iOS-SDK

**Comment by kevjose** [comment]
@ben-xD I can confirm the above, I used a service from Sendbird which sends a push notification via APNs directly and your solution does work. However, when I try with either OneSignal or even Firebase Messaging it does not work (meaning the onMessage is never invoked).

**Comment by kevjose** [comment]
On a separate note:
@ben-xD how do i override the `UserNotificationCenterDelegateHandlers.userNotificationCenter` implementation via AppDelegate.swift. 

Currently, if I define an `override func userNotificationCenter` in the AppDelegate.swift, it picks up the one defined in the `push` library (ios/.symlinks/plugins/push_ios/ios/Classes/UserNotificationCenterDelegateHandlers.swift)

**Comment by ben-xD** [comment]
If you're using Firebase Messaging on iOS, you shouldn't use this package. This package specifically exists for users who don't want to use Firebase on iOS.

I'm not sure what you mean about overriding an implementation class of this plugin. You shouldn't override it if you use this package. 

The original delegate should be called when necessary. So if you set your own delegate, that should still work.

The problem I can see is other iOS push notification packages perform swizzling or replacing delegates, breaking support for other packages. Because you can only have 1 delegate (effectively the delegate is a singleton), plugin developers should be aware to call the original delegate (I do that).

I don't think there is anything I can do if Firebase Messaging or OneSignal breaks the system for other plugins. :)

**Comment by kevjose** [comment]
Thanks for the reply @ben-xD. This does make sense to me 👍

**Comment by kaanbasal** [comment]
@ben-xD I think this issue is still relevant. after checking the code in this repo and debugging the application on iOS, if I am not mistaken, you are using `silent push notifications` aka `content-available` to test all of these. but as per apple's documentation, we shouldn't be using `silent push notifications` more than 2/3 times per hour. 
see: https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/pushing_background_updates_to_your_app

what I am not sure is: if we add "alert" in the APNS payload, does this message still count as `silent push notification` or not.

I think, we should be able to receive PN and do whatever we want to do with it, eg: show toast message.

https://github.com/ben-xD/push/blob/dd0e4972aa44fc02368c2a4a819ca45d255f7771/push_ios/ios/Classes/UserNotificationCenterDelegateHandlers.swift#L13-L28

with the lines above, it is not possible for us to capture that information at all.

otherwise, can you please let me know how to show `Toast Message` when the application gets push notification while its in foreground, for both ios & android.

Thank you

**Comment by ben-xD** [comment]
Hey @kaanbasal,

> what I am not sure is: if we add "alert" in the APNS payload, does this message still count as silent push notification or not.

It depends on the header of the request sent to APNs. When you use `"apns-push-type: alert"`, it is not a background push notification (which would be `"apns-push-type: background"`). So that documentation you linked doesn't apply.

> I think, we should be able to receive PN and do whatever we want to do with it, eg: show toast message.

You **should** receive if you use this app, on the `onMessage` handler. If you have OneSignal or Firebase Messaging installed, it might not work if those packages do unexpected changes to the application (e.g. swizzling methods or replacing delegates).

> with the lines above, it is not possible for us to capture that information at all.

**In what cases does it not work for you?** I can improve that section of code but I'm not sure what problems you're facing.

> otherwise, can you please let me know how to show Toast Message when the application gets push notification while its in foreground, for both ios & android.

Listen to the `onMessage` stream and show a toast. It shouldn't need more than that :)

**Comment by kaanbasal** [comment]
Thanks for the reply @ben-xD ,

> So that documentation you linked doesn't apply

Thank you for the clarification, in that case I can send every push notification with `content-available:1` and library should be working without a problem.

> If you have OneSignal or Firebase Messaging installed

I am not using any other library as they are either too heavy or depends on other services. 

> I'm not sure what problems you're facing

I tried listening `onMessage` stream when application is on the `foreground` without setting `content-available:1`. this was the only case I couldn't get any callback from library. if I change push payload using `content-available:1` then it is being called.

So based on these informations my questions would be: 
- Why do we need to set `content-available:1` just to be able to handle notifications when app is on foreground? or is there any other use case of this `content-available` flag that I am missing?
- Does the change I suggested above makes sense in this case?

**Comment by ben-xD** [comment]
Hey @kaanbasal, thanks for your message and your time in explaining.

> Why do we need to set content-available:1 just to be able to handle notifications when app is on foreground? or is there any other use case of this content-available flag that I am missing?

Oh that's an interesting observation. So the message I've been testing is both an alert and a background message. And perhaps there are limitations on how many messages with `content-available` will be received. So we shouldn't have this constraint.

> Does the change I suggested above makes sense in this case?

I've made some changes in https://github.com/ben-xD/push/commit/5bae1053bb146008e6d284ebe36987300ef2c610, which takes your suggestion into consideration. I had to add a few more bits to avoid duplicate notifications in some cases. I've also tested the case for app in foreground and background, and with and without `content-available = 1`.

Let me know how it goes :)

**Comment by kaanbasal** [comment]
Hey @ben-xD thank you for the fast reply, the fix commit and also for publishing :) 

I just had time to test it out and it seems its working as expected when app is on the foreground. Tested without setting `content-available` 

But please be aware that I tested only the part I mentioned and I don't know if there are any side effects of this change.

I will be working on push notifications soon and I will let you know if anything goes wrong.

Until then take care

**Comment by ben-xD** [comment]
With [`UNNotificationServiceExtension`](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension), I'm not sure if `push` needs to use this, given that you could create your own local notification using `flutter_local_notifications`? It doesn't look like it supports images though.

OneSignal's documentation for setting it up is [Step 3. Add an iOS Service Extension (iOS Apps Only)](https://documentation.onesignal.com/docs/flutter-sdk-setup#step-3-add-an-ios-service-extension-ios-apps-only).
```
    /* DEBUGGING: Uncomment the 2 lines below and comment out the one above to ensure this extension is excuting
                  Note, this extension only runs when mutable-content is set
                  Setting an attachment or action buttons automatically adds this */
```

What use case are you looking at that you need `UNNotificationServiceExtension`?
